### ISSUE TITLE
Use a decent code generator or templates

### ISSUE THREAD

**Issue by mweststrate** [mobxjs/mst-gql#13]
The current generator `generate.js`, is a pretty straight forward string concatenation tool. it's fast and straight forward, but probably not very maintainable as it grows more complex in the future. Something more solid could be a good idea, I think @birkir had some thoughts about that in the past

**Comment by mweststrate** [comment]
For example: https://github.com/dotansimha/graphql-code-generator

**Comment by sw-yx** [comment]
@jamonholmgren's https://github.com/infinitered/gluegun goes templating well.

**Comment by jamonholmgren** [comment]
Thanks for the mention, @sw-yx!

We're using the `ejs` library to power our template system:

https://github.com/infinitered/gluegun/blob/b35a848c1cf5324f7767d15104416d8a78b2035b/src/toolbox/template-tools.ts#L16

We also have a really cool patching library:

https://github.com/infinitered/gluegun/blob/b35a848c1cf5324f7767d15104416d8a78b2035b/src/toolbox/patching-tools.ts

Both are built-in to Gluegun, which may be overpowered for your needs here. But if you want to make `mst-gql` a full featured CLI, let me know and I can walk you through what you get with Gluegun! Otherwise, feel free to steal/borrow any concepts from the links I posted here.

**Comment by FredyC** [comment]
Should we actually turn this into a monorepo to split stuff into plugins? I am no big fan of Lerna, it does the job, but it also gets in the way fairly often.

**Comment by mweststrate** [comment]
Let's not over engineer yet; it does make contributing harder

Op wo 29 mei 2019 11:10 schreef Daniel K. <notifications@github.com>:

**Comment by FredyC** [comment]
Ok sure, I suppose that Gluegun can use "plugins" from the single package, so that shouldn't matter, but we should also support custom plugins from NPM.

---

Regarding the code generation, I actually do have a custom generator for `graphql-code-generator` which generates fully typed hooks for each graphql operation. I am not sure yet how it would like here, but a similar alternative would be really nice.

```ts
export function useQResolveLocation(
  variables?: QResolveLocationVariables,
  baseOptions?: Hooks.QueryHookOptions<QResolveLocationVariables>,
) {
  return Hooks.useQuery<QResolveLocationQuery, QResolveLocationVariables>(
    QResolveLocationDocument,
    variables,
    baseOptions,
  )
}
```

**Comment by mweststrate** [comment]
Please open separate issues for separate ideas 😉

Op wo 29 mei 2019 11:58 schreef Daniel K. <notifications@github.com>:

**Comment by dpnolte** [comment]
I just want to add that kotlin/java poet might be a good source of inspiration: https://github.com/square/javapoet and https://github.com/square/kotlinpoet . I love that library. It uses builders to generate the code, instead of dealing with a template engine. The builders take care of indentation, importing types and injecting formatted variables with c-style literals. Writing code generator logic can get quite messy quickly, and java poet does a great job of keeping it under control. Unfortunately, I didn't find any equivalents in JS. The most similar lib that I've found so far was [ts-morph](https://github.com/dsherret/ts-morph)

**Comment by RXminuS** [comment]
Would love to help out with this.

Gluegun definitely looks promising (especially the patching could be great). I've also used [Plop](https://github.com/amwmedia/plop) which was quite simple to understand and [Hygen](https://github.com/jondot/hygen) looks promising.

@dpnolte ts-morph looks like "the proper way" that you would want to do code generation by using the AST. However, reasoning about AST's gets quite complex quickly and can be difficult for newcomers to understand. Since most of the code will be auto-generated and provide convenient hooks for people to extend it means we already have a great deal of control over some of the foundational assumptions and as such we could just work on those with simpler file templating and "patch comments" for us to use. 

I think we should make a decision on this rather soon before we end up adding a bunch of code that becomes difficult to port. @mweststrate did you already have a direction in mind?

**Comment by mweststrate** [comment]
Nope! I think the person giving this a shot is in the best position to
judge :). But yeah, I agree that the AST direction is most probably adding
unnecessary burden

**Comment by RXminuS** [comment]
Ok, but then I'll give it a shot with Gluegun next week. I'll probably first just spit out a few ejs templates not worrying too much about the actual data. And then put up a bit cleaner framework for the CLI & logging. Hopefully, then others can join in to help fill in the actual template data as I expect that requires a bit of knowledge of how it is currently generated.

I would prefer writing everything in Typescript if that's ok with everyone?

**Comment by chrisdrackett** [comment]
yes please on typescript :)

**Comment by jamonholmgren** [comment]
@RXminuS `npx gluegun new mst-gql --typescript` will get you set up with an initial TypeScript boilerplate. Hit me up if you run into problems. I'm happy to knock out bugs fast on Gluegun. I just released 4.0.0 which needs more testing, so let me know. Also any documentation gaps.

**Comment by beepsoft** [comment]
@RXminuS do you have any progress with this?

I myself would like to have the options provided by https://graphql-code-generator.com/docs/plugins/typescript, which allow changing the case of type names/enums and getting rid of underscores in type names in generated code. 

It uses a configuration like this:

```
config:
  namingConvention:
    typeNames: change-case#pascalCase
    transformUnderscore: true
```

I thought I add such options to the current generator but if there's already a better one in progress I would rather work with that.
